code/modules/library/book.dm

```diff
@@ -0,0 +1,370 @@
+		return
+	read_book(user)
+
+/obj/item/book/attackby(obj/item/I as obj, mob/user as mob, params)
+	if(istype(I, /obj/item/pen))
+		edit_book()
+	else if(istype(I, /obj/item/barcodescanner))
+		var/obj/item/barcodescanner/scanner = I
+		scanner.scanBook(src, user) //abstraction and proper scoping ftw | did you know barcode scanner code used to be here?
+		return
+	else if(I.sharp && !carved) //don't use sharp objects on your books if you don't want to carve out all of its pages kids!
+		carve_book(user, I)
+	else
+		store_item(I)
+
+/**
+  * Internal Checker Proc
+  *
+  * Gives free pass to observers to read, ensures that all other mobs attempting to read book are A) literated and
+  * B) are within range to actually read the book.
+  */
+/obj/item/book/proc/can_read(mob/user)
+	if(isobserver(user)) //We check this first because ghosts should be able to read any book they can see no matter what
+		return TRUE
+	if(!in_range(src, user))
+		return FALSE
+	if(!user.is_literate()) //this person was 2 cool 4 school and cannot READ
+		user.visible_message("You attempt to the read the book but remember that you don't actually know how to read.")
+		return FALSE
+	return TRUE
+
+/**
+  * Read Book Proc
+  *
+  * Checks if players is able to read book and that book is readable before calling the neccesary procs to open up UI
+  */
+/obj/item/book/proc/read_book(mob/user)
+	if(!length(pages)) //You can't read a book with no pages in it
+		to_chat(user, "This book is completely blank!")
+		return
+	if(!can_read(user))
+		return
+	user.visible_message("[user] opens a book titled \"[title]\" and begins reading intently.")
+	show_content(user) //where all the magic happens
+	onclose(user, "book")
+
+/**
+  * Show Content Proc
+  *
+  * Builds the browserUI html to show to the player then open up the User interface. It first build the header navigation
+  * buttons and then builds the rest of the UI based on what page the player is turned to.
+  */
+/obj/item/book/proc/show_content(mob/user)
+	var/dat = " "
+	//First, we're going to choose/generate our header buttons for switching pages and store it in var/dat
+	if(length(pages)) //No need to have page switching buttons if there's no pages
+		if(current_page == 0)
+			dat+= "<DIV STYLE='float:left; text-align:left; width:49.9%'></DIV>"
+			dat+= "<DIV STYLE='float:left; text-align:right; width:49.9%'><A href='?src=[UID()];next_page=1'>Next Page</A></DIV><BR><HR>"
+		else if(current_page < length(pages))
+			dat+= "<DIV STYLE='float:left; text-align:left; width:49.9%'><A href='?src=[UID()];prev_page=1'>Previous Page</A></DIV>"
+			dat+= "<DIV STYLE='float:left; text-align:right; width:49.9%'><A href='?src=[UID()];next_page=1'>Next Page</A></DIV><BR><HR>"
+		else if(current_page == length(pages))
+			dat+= "<DIV STYLE='float:left; text-align:left; width:49.9%'><A href='?src=[UID()];prev_page=1'>Previous Page</A></DIV>"
+			dat+= "<DIV STYLE='float;left; text-align:right; width:49.9%'></DIV>"
```

Comment:
Prob can be made a bit cleaner by having a var for each div which you insert
something like:
```dm
var/left
var/right
if(cur_page == 0)
 right = ...
if(cur_page < len(pages)
 right = ...
 left = ...
...
date += "<div...>[left]</div>"
date += "<div...>[right]</div>"
```
Also HTML elements are written in lowercase normally. So `<div>` instead of `<DIV>`

Reply:
cleaned up